canonical/cloud-init
# feat(WSL): Add support for Ubuntu Pro configs

## Proposed Commit Message

```
feat(WSL): Add support for Ubuntu Pro configs

Currently, WSL only supports manual cloud-init configurations provided
in the host Windows filesystem. This adds support for Landscape/Ubuntu
Pro for WSL to provide cloud-init configurations and have them merged 
with or override manual user configurations. This adds support for 
organizations to better provision WSL instances using cloud-init.
```

## Additional Context
Currently, the WSL datasource can only work with user-provided configurations manually created in the host Windows filesystem. This change allows organizations to provide configurations via Landscape/Ubuntu Pro and have them loaded into WSL instances. These new configurations will be created under `%USERPROFILE%/.ubuntupro/.cloud-init`. This change will require [ubuntu-pro-for-wsl](https://github.com/canonical/ubuntu-pro-for-wsl) to implement these files after this is merged, as already created in https://github.com/canonical/ubuntu-pro-for-wsl/pull/506.

## Test Steps

As described in the updated docs and the spec sheet for this change, you can test this by manually creating the necessary files on your Windows host system. If an `agent.yaml` file is found, it should properly merge (overriding any conflicts) on a per-module basis with either a Landscape config (in `%USERPROFILE%/.ubuntupro/.cloud-init/<InstanceName>.user-data`) or a user-data found by the previous implementation as described in #4786.

ds-check should also verify the environment by using any one of the files described above.

## Checklist


## Merge type


Edit:
*Force pushed to add more context to some commits*

UDENG-2381
